### PR TITLE
[WIP] Map.get with projection offloading

### PR DESCRIPTION
**Pull request by tombujok**

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/GetWithProjectionOperation.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    public GetWithProjectionOperation(String name, Data dataKey, Projection projection) {
+        super(name, dataKey);
+
+        this.dataKey = dataKey;
+        this.projection = projection;
+    }
+
+    @Override
+    public void run() {
+        Object value = recordStore.get(dataKey, false);
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/GetWithProjectionOperation.java]: You might run into a data race here if the object is stored in OBJECT format. If you don't clone, the projection thread(s)! and partition thread (modifying EP) could be accessing the same object instance,

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/GetWithProjectionOperation.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        runOffloadedProjection(value, (OperationServiceImpl) getNodeEngine().getOperationService());
+    }
+
+    private void runOffloadedProjection(final Object value, final OperationServiceImpl ops) {
+        ops.onStartAsyncOperation(this);
+        getNodeEngine().getExecutionService().execute(ASYNC_EXECUTOR, new Runnable() {
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/GetWithProjectionOperation.java]: For the time being it is ok, but in the future having control on the executor is the better choice.

File: hazelcast/src/main/java/com/hazelcast/core/IMap.java [hunk 3]
```diff
@@ -80,6 +80,18 @@
...
+     * Gets the object stored under the given key, applies the given projection to it and returns output of the projection.
+     * The object fetching will happen on the partition-thread, the projection will be applied on a generic thread.
+     * Return a clone of the value.
+     *
+     * @param k   key to fetch the value from
+     * @param p   projection to apply
+     * @param <O> type of the output object
+     * @return result of the projection that is applied to the object fetch from the given k.
+     */
+    <O> O get(Object k, Projection<V, O> p);
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/core/IMap.java]: Should k be of type object or K?

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/GetWithProjectionOperation.java (hunk 1 above)
Comment (tombujok) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/GetWithProjectionOperation.java]: Thx. fixed.

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/GetWithProjectionOperation.java (hunk 2 above)
Comment (tombujok) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/GetWithProjectionOperation.java]: Added GroupProperties

File: hazelcast/src/main/java/com/hazelcast/core/IMap.java (hunk 3 above)
Comment (tombujok) [inline hazelcast/src/main/java/com/hazelcast/core/IMap.java]: Map's contract somehow defines it as Object even though there's K

**Comment by tombujok** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/GetWithProjectionOperation.java (hunk 2 above)
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/GetWithProjectionOperation.java]: Eventually I would coordinate this on a call level. So each call should be able to configure the executor it runs on. This way the caller has fine grained control on what gets run in parallel and what gets queued.If you have a single global property; the caller has not a lot of control.

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/GetWithProjectionOperation.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    @Override
+    public void run() {
+        Object value = recordStore.get(dataKey, false);
+
+        boolean offload = getNodeEngine().getProperties().getBoolean(GET_PROJECTION_OFFLOAD);
+        SerializationService ss = getNodeEngine().getSerializationService();
+        if (offload) {
+            value = isObjectFormat() ? ss.toObject(ss.toData(value)) : ss.toObject(value);
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/GetWithProjectionOperation.java]: I would move the toObject into the offloading. So you convert object -> data on the partition thread, but the offloading thread can take care of deserializing data to object. This removes more pressure for the partition thread and will allow for more parallel operations on the same partition.

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/GetWithProjectionOperation.java (hunk 2 above)
Comment (tombujok) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/GetWithProjectionOperation.java]: Sure, the GroupProperty was added for the sake of PoC for now

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/GetWithProjectionOperation.java (hunk 4 above)
Comment (tombujok) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/GetWithProjectionOperation.java]: Fixed

File: hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientMapProxy.java [hunk 5]
```diff
@@ -286,14 +288,25 @@ public V get(Object key) {
...
+        Data projectionData = toData(projection);
+        return toObject(getInternalWithProjection(keyData, projectionData));
     }
 
-    protected Object getInternal(Data keyData) {
-        ClientMessage request = MapGetCodec.encodeRequest(name, keyData, getThreadId());
+    protected Object getInternalWithProjection(Data keyData, Data projection) {
+        ClientMessage request = MapGetWithProjectionCodec.encodeRequest(name, keyData, projection, getThreadId());
         ClientMessage response = invoke(request, keyData);
         MapGetCodec.ResponseParameters resultParameters = MapGetCodec.decodeResponse(response);
```
Comment (sancar) [inline hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientMapProxy.java]: Need to replace with
`MapGetWithProjectionCodec.ResponseParameters resultParameters = MapGetWithProjectionCodec.decodeResponse(response);`

Comment (tombujok) [inline hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientMapProxy.java]: Thx @sancar Fixed!

**Pull request closed without merging**